### ISSUE TITLE
Memory debugging are missing lots of features from Observatory

### ISSUE THREAD

**Issue by jensjoha** [flutter/devtools#5761]
The "Allocation Profile" option on Observatory is, I think, being replaced by "Memory" in devtools.
Right of the bat I am missing details about new and old heap, and internal vs external memory. I seemingly have to download the CSV file to get that info.
Next of I can no longer press any of the classes to get a detailed view (let alone go from there).

I answered something similar in the questionnaire that was given out some time ago (might have been years at this point, I'm not sure), but I'll repeat it here.

When debugging memory issues I often do this:

- Open observatory.
- Go to "Allocation Profile" and find the class. Click it.
- Use one of the "strongly reachable", "all direct instances", "all instances of subclasses" or "all instances of implementors" options.
- Click the resulting list, either going through it by hand (again clicking on the actual instances etc) and/or performing expression evaluations on the list.
- Open two or more instances in a new tab so I can visually compare them by tabbing back-and-forth.
- Use the "Retaining path" and "Inbound references" feature on instances.

Seemingly _none_ of this is possible though devtools. (and if it is I can't find it).

More advanced I've also hacked together scripts that could find interesting instances, I could then print their id and go to observatory for a visual of this instance (and again for instance use the  "Retaining path" features etc) by changing the url for another instance (changing the (e.g.) "&objectId=objects%2F6661" part).

Can I please have this functionality back?

/cc @bkonyi @kenzieschmoll

**Comment by polina-c** [comment]
@jensjoha , feel free to schedule meeting with me and @bkonyi  discuss what is missing. We will either point you to new way of doing what you want or create separate issues to fill the gaps.

**Comment by jensjoha** [comment]
So the developer mode does give the new space old space stuff.

The "Objects" tool under "VM Tools" isn't much use --- the best I can tell I have to navigate to "Class Hierarchy" and navigate in the non-searchable list to the class I want, then I can click on it, but can't get on from there (get a list and perform expression evaluation on it).

I see no need for scheduling a meeting. What is missing is the possibility go get all instances of a specific class (i.e. get an instance of a list with all instances in it), see details about instances and perform expression evaluation on instances. (Preferably it should be easier to get there than by having to mouse-navigate a mile-long nested list where one has to know the class inheritance hierarchy by heart, there shouldn't be text on top of other text (image 1) and scroll bars shouldn't be missing (image 2)).

Image 1: Text on top of each other.
![Image 1](https://user-images.githubusercontent.com/20773519/237015945-54973356-c154-4e5b-b1ec-af2bf1f8f4c9.png)

Image 2: No scrollbar and can't scroll despite there being more data.
![Image 2](https://user-images.githubusercontent.com/20773519/237015980-ac9b9391-4041-4ac1-8f4e-883f967caf58.png)

**Comment by polina-c** [comment]
You may want to check this section: https://docs.flutter.dev/tools/devtools/console#browse-heap-snapshot

**Comment by jensjoha** [comment]
This, as I read it, requires one to take a heap snapshot which - with even a rather small instance running - takes minutes and uses 4+GB of ram before crashing the tab in chrome.

**Comment by bkonyi** [comment]
Yes, that does require use of a heap snapshot. @polina-c, we likely need to be able to support a similar workflow that relies on live instance evaluation if we're hitting memory limits in the browser.

**Comment by polina-c** [comment]
Created separate issue: https://github.com/flutter/devtools/issues/5776

**Comment by kenzieschmoll** [comment]
@bkonyi have all the issues reported here been resolved?

**Comment by bkonyi** [comment]
AFAIK, yes, but @jensjoha should confirm before we close this.

**Comment by jensjoha** [comment]
Today I needed to find the 1 live instance of a specific class and find out why it was still alive.

In observatory I did this:
* click "allocation profile"
* Ctrl+F and search for my class. I write `synthetictoken` and it's found. I click it.
* I scroll down to "all direct instances", click the thing, extend the list, click the 1 result.
* Under "Retaining path" I click the thing and I get what I want.
* (Additional note: Had there been two instances I would have held ctrl while clicking on each result, thus opening a new tab for each, and for each tab click "Retaining path".)

I then try the same thing in Dart DevTools:
* Click "memory"
* Ctrl+F, the browser default search pops up and doesn't work.
* I look around, find the filter-icon and use my mouse to click it.
* A big thing pops up --- I use my mouse to click "show only", write `synthetictoken`, hit enter. I get a new line. Backspace. Use my mouse to click apply. Nothing is found.
* I make a guess that this is because of casing, redo the thing, write `SyntheticToken`, use my mouse to click apply. I found my result.
* Now what? I click the line and after a while I notice 3 small dots next to the "1" in "Instances". I click it. I can select "store all instance" -> "all direct instances". After a while a list pops up in the console. I can extend the list and see my one item.
* Now what? Can I just not find the "Retaining path" button, or is it missing? I give up.
* (For the fun of it I want to do an expression compilation on the list --- but I seemingly can't. I type `?`, read about `$0` etc -- try it out, but that doesn't work either.)

So while I can theoretically get a list of "all direct instances" etc (one of the requests) I still can't expression evaluate on it, easily open new tabs for new things (e.g. for each instance to do expression evaluation on separately) or get the retaining path and/or inbound references.

**Comment by bkonyi** [comment]
Thanks for the feedback Jens!

I tried to reproduce a similar flow and also encountered difficulty.

There's a few actionable items I can see, some that should be simple QoL fixes and a couple bigger gaps:

- We should have better keyboard shortcut support for things like CTRL+F, Return to confirm / close a dialog, etc.
- We should consider making search boxes and filters case insensitive (at least by default)
- The selector to move instances to the evaluation console isn't very discoverable and should be made more clear.
- There's no way to retrieve the retaining path of a live object through the console. We should consider adding a console command to allow for this.
- The expression evaluation console needs some focus during the bug bash as sometimes it doesn't work as expected (e.g., `var y = $0` was giving a compilation error, but it's not clear how it gets into this state)

After playing with things a bit more, I was able to find retaining path information through the `Diff Snapshots` tab by collecting a snapshot and filtering for the type I'm interested in. Selecting the type brings up a list of retaining paths. 

It looks like this flow needs some work to make the retaining path more clear (there's no way to tell what script, class, or library is on the path). Can you give this a try and let us know if there's anything else in particular that might be missing?

<img width="1385" alt="image" src="https://github.com/flutter/devtools/assets/24210656/ec65ec3c-a789-470d-b353-6c817db20ff1">

There's also support for inspecting the retaining path of an individual object under the `VM Tools -> Objects` tab, but this only works if the object you want to inspect is a static variable. Even then, you need to manually navigate to the owning library / class to get access to the field to even inspect its static value. I'd like to provide a way for users to more easily inspect individual instances in this view, but it'll require some UX work to figure out.

<img width="1399" alt="image" src="https://github.com/flutter/devtools/assets/24210656/82b06c5f-fbb7-4f03-bf44-fb47daaabca3">

@polina-c, can you please let me know if I glanced over any functionality that would help Jens with this particular workflow?

**Comment by polina-c** [comment]
@jensjoha , if you just take heap snapshot, you will see retaining pathes for each instance. 
It is at Memory > Diff Snapshot

Let me know if it helps.

**Comment by jensjoha** [comment]
As in https://github.com/flutter/devtools/issues/5761#issuecomment-1541415871 taking a memory snapshot doesn't work. It takes forever and then the tab crashes (yes, I tried again --- and this is with the process using <1GB of memory and the isolate it should be taking a snapshot of using <500MB --- for reference I find myself needing to get the live objects of a specific type and getting the retaining path etc on processes using 4-10 GB quite often).

As for adding a console command to get the path for a live object, that would certainly be better than nothing, but I'd at least like it to take up quite a bit more screen real-estate than the console does (I realize it can be resized but that will get real annoying real fast).

And again I'd like to stress the point about being able to open tabs for each object. I rarely only have one object to look at. Often I look at many. Sometimes I visually compare them by tabbing back-and-forth etc.
